engine/pkg/election/elector.go

```diff
@@ -0,0 +1,353 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package election
+
+import (
+	"context"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+func init() {
+	rand.Seed(time.Now().UnixNano())
+}
+
+// Elector is a leader election client.
+type Elector struct {
+	config Config
+
+	observeLock    sync.RWMutex
+	observedRecord Record
+	// observedRenews is a map of renew time of each member.
+	// Note that the time is not RenewTime recorded in the record,
+	// but the time when we observed the renewal. This is because
+	// absolute time is not reliable across different machines.
+	observedRenews map[string]time.Time
+
+	// resignCh is used to notify the elector to resign leadership.
+	// The value is the duration that the elector should not be leader.
+	resignCh chan time.Duration
+	// Elector will not be leader until this time.
+	resignUntil time.Time
+}
+
+// NewElector creates a new Elector.
+func NewElector(config Config) (*Elector, error) {
+	if err := config.AdjustAndValidate(); err != nil {
+		return nil, err
+	}
+	return &Elector{
+		config:         config,
+		observedRenews: make(map[string]time.Time),
+		resignCh:       make(chan time.Duration),
```

Comment:
Do we need to add one buffer for `resignCh`, and add a default branch at L350?
In some corner case, election could stuck in a deadloop of `renew`, which leads to no channel receiver of `resignCh` and block of `ResignLeader` API.

Reply:
There is a timeout for `renew` loop. So this API shouldn't be blocked for too long. No buffer is intended to make the resign duration more intuitive (It's calculated after API response).

Reply:
But there is indeed a bug here. When `elector` is closed, `ResignLeader` will be blocked forever. Maybe we should add a timeout here.